File: app/autopos/item/service/ItemService.scala
package autopos.item.service

import javax.inject.{Inject, Singleton}

import autopos.item.model.{Item, ItemCommand, ItemDto}
import autopos.item.service.repo.ItemRepo
import autopos.shared.structure.{Page, Pageable}
import com.google.inject.ImplementedBy

import scala.concurrent.Future

@ImplementedBy(classOf[ItemServiceImpl])
trait ItemService {

  def addItem(itemCommand: ItemCommand): Future[Long]

  def updateItem(id: Long, itemCommand: ItemCommand): Future[Int]

  def getItems(pageable: Pageable): Future[Page[ItemDto]]

  def getItem(id: Long): Future[Option[ItemDto]]

  def deleteItem(id: Long): Future[Int]

}

@Singleton
class ItemServiceImpl @Inject()(itemRepo: ItemRepo)
  extends ItemService {

  override def getItems(pageable: Pageable) =
    itemRepo.list(pageable)

  override def getItem(id: Long) =
    itemRepo.findById(id)

  override def updateItem(id: Long, itemCommand: ItemCommand) =
    itemRepo.update(id, itemCommand)

  override def addItem(itemCommand: ItemCommand) = {
    itemRepo.create(itemCommand)
  }

  override def deleteItem(id: Long) =
    itemRepo.delete(id)

}

File: ui/src/app/catalog/items/item_service.ts
import {Http} from 'angular2/http';
import {Injectable} from 'angular2/core';
import {Pageable} from '../../shared/Pageable';
import {ItemCommand} from './item_command';

@Injectable()
export class ItemService {

  private baseUrl = "http://localhost:9000/api/items";

  constructor(private http:Http) {
  }

  public getItems(pageable:Pageable) {
    let url = this.baseUrl + "?" + pageable.queryString();
    return this.http.get(url);
  }

  public getItem(id:number) {
    let url = this.urlWithId(id);
    return this.http.get(url);
  }

  public addItem(itemCommand:ItemCommand) {
    return this.http.post(this.baseUrl, JSON.stringify(itemCommand));
  }

  public updateItem(itemCommand:ItemCommand) {
    let url = this.urlWithId(itemCommand.id);
    return this.http.put(url, JSON.stringify(itemCommand));
  }

  public deleteItem(id:number) {
    let url = this.urlWithId(id);
    return this.http.delete(url);
  }

  private urlWithId(id:number) {
    return this.baseUrl + "/" + id;
  }
}

